src/Chainweb/Chainweb/PruneChainDatabase.hs

```diff
@@ -0,0 +1,438 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- |
+-- Module: Chainweb.Chainweb.PruneChainDatabase
+-- Copyright: Copyright © 2020 Kadena LLC.
+-- License: MIT
+-- Maintainer: Lars Kuhtz <lars@kadena.io>
+-- Stability: experimental
+--
+-- There is a block header database for each, but only a single shared payload
+-- database.
+--
+-- Since BlockTransactions can be shared between blocks, even across chains,
+-- we'll have to take a global approach at garabage collecting payloads.
+--
+-- There's also the advantage of guaranteeing that there are no concurrent
+-- writes during the garbage collections, because the chain databases aren't
+-- initialized yet.
+--
+-- 1. Concurrently for each chain:
+--     * initialize temporary block header db,
+--     * prune the block header db, and
+--     * use the callbacks to delete those parts of the payloads that are
```

Comment:
This needs an update re payload deletion no?

Reply:
yes, the comment is outdated.

Reply:
Actually, a lot of the module comment is outdated.

Reply:
There are two pruning modes:

1. `forks`. Only remove block headers that are not reachable from the most recent blocks of the chain. Doesn't touch payloads at all.

2. `full`. Does header plus mark and sweep GC for payloads. GC takes a conservative approach that guarantees that even if something goes wrong in between, there are no dangling references in the -- even for non-reachable entries.

   1. mark all used payload hashes during block header pruning
   2. traverse all `BlockPaylaods` and sweep all unmarked payload entries and mark all transactions hashes and outputs hashes that are used by marked `BlockPayload`s.
   3. concurrently sweep all unmarked `BlockTransactions` and `BlockOutputs`.

So, it is guaranteed that if if a BlockPayload is kept (is marked) also the respective transactions and outputs are marked and kept. The algorithm works top-down, so referenced entries are deleted only after the referencing entry is deleted. If an exception occurs during the GC, some garbage is remains but there are not dangling references, i.e. the database is consistent.